articles/115_topic_and_service_name_mapping.md

```diff
@@ -0,0 +1,317 @@
+A relative name does not begin with a forward slash (`/`) and does respect the namespace of the node which created them.
+
+Relative names are appended to the namespace of the node which creates them.
+For example, if the node is put into a namespace `/ping/pong` and the topic or service name is `foo/bar`, then the absolute name will become `/ping/pong/foo/bar`.
+However, if the topic or service name is `/foo/bar`, then the absolute name will stay just `/foo/bar`, ignoring the node's namespace.
+
+### Name Tokens
+
+Name token's are the strings between the namespace delimiters, e.g. the tokens of the topic or service `/foo/bar/baz` are `foo`, `bar`, and `baz`.
+
+Topic and service name tokens:
+
+- must not be empty, e.g. the name `//bar` is not allowed
+
+  - rational: it removes the chance for accidental `//` from concatenation; also if the implementation chooses to use a trailing underscore (`_`) then it prevents `foo_/_bar` and `foo//bar` from being ambiguous
+
+- may use alphanumeric characters (`[0-9|a-z|A-Z]`), or an underscore (`_`)
+
+- must not start with numeric characters (`[0-9]`)
+
+- must not end with a single underscore (`_`)
+
+  - rational: if tokens are allowed to end with and start with `_` then `foo_/bar` is indistinguishable from `foo/_bar` if the `/` is replaced with `__`, i.e. both result in `foo___bar`
+
+- must not be a single underscore (`_`)
+
+  - rational: this is a special case of "must not end with an underscore"
+
+- must not have two or more underscores (`__`) repeated anywhere
+
+  - rational: this provides the implementation with a safe to use delimiter
+
+- may be a single tilde character (`~`)
+
+### Private Namespace Substitution Character
+
+The special single character token `~` will be replaced with a namespace snippet that is a concatenation of the namespace for the node and the node name.
+For example, for a node `node1` in a namespace `/foo` would result in `~` being replaced with `/foo/node1`.
+For another example, a node `node1` in a namespace `foo/bar` would result in `~` being replaced with `foo/bar/node1`.
+It may be used in any token position.
+Any resulting empty namespaces which result from expanding the `~` will be collapsed.
+For example, when `~` expands to `/foo/node1` and the name `ping/~/pong` is given, it will expand to `ping/foo/node1/pong` and not to `ping//foo/node1/pong`.
+Here is a table with some example expansions:
+
+| **Input Name** | Node: `my_node` NS: none | Node: `my_node` NS: `/foo`     |
+|----------------|--------------------------|--------------------------------|
+| `ping`         | *`/ping`*                | *`/foo/ping`*                  |
+| `/ping`        | *`/ping`*                | *`/ping`*                      |
+| `~`            | *`/my_node`*             | *`/foo/foo/my_node`*           |
+| `~/ping`       | *`/my_node/ping`*        | *`/foo/foo/my_node/ping`*      |
```

Comment:
Yeah, I was imagining something like a topic called `/public_cameras/~/image_raw` and then two node instances per robot, where the node names are `left_camera` and `right_camera`, and each robot is in a namespace like `robot1` and `robot2`. You'd get something like this:

- `/public_cameras/robot1/left_camera/image_raw`
- `/public_cameras/robot1/right_camera/image_raw`
- `/public_cameras/robot2/left_camera/image_raw`
- `/public_cameras/robot2/right_camera/image_raw`

Now, I don't know why you'd want to do that, and you could do it in ROS 1, but not in a way that's easily remapped. In this scenario you can do a remap like `/public_cameras/~/image_raw:=/private_cameras/~/image_raw`. However, without this more flexible `~`, you'd have to write Python code like this to achieve it: `pub = Publisher('/public_camera/' + rospy.get_namespace() + '/' + rospy.get_name() + '/image_raw', ...)`. That's fine, except the remapping becomes non-trivial, something like `/public_camera/robot1/left_camera/image_raw:=/private_camera/robot1/left_camera/image_raw` where the `robot1/left_camera` part would be subject to change based on the configuration.

---

So that's what I was imagining, but I'm also ok with keeping it simple. What do others think?